### DISCUSSION TITLE
[@types/react] Some intrinsicElements are off

### DISCUSSION THREAD

**Discussion by sonicolasj** [DefinitelyTyped/DefinitelyTyped#58891] [Issues with a @types package] [answered]
Hi!

## Analysis

IntrinsicElements are all declared as `React.DetailedHTMLProps<A<B>, C>`, where the basic version of `A` is `React.HTMLAttributes`, and the basic version of both `B` and `C` is `HTMLElement`. 

In this case, the `A` type determines the attributes available to use in React, for the given tag, and both `B` and `C` determines the type of the `e.currentTarget` in the event listeners add to the tag.

So, a tag implementing HTMLElement and only having globally defined attributes is `React.DetailedHTMLProps<React.HTMLAttributes<HTMLElement>, HTMLElement>`.

## Issue

There is a handful of elements, while having more attributes than the globally defined ones, are typed as `HTMLElement` in their event handlers callbacks, losing their particular attributes without manual retyping.

Example with `<details>`:

```tsx
// for reference
interface IntrinsicElements {
  details: React.DetailedHTMLProps<React.DetailsHTMLAttributes<HTMLElement>, HTMLElement>;
}

// Without explicit typing: causes an error.
const TestComponent1 = () => (
  <details open={false} onToggle={(e) => console.log('is now open?', e.currentTarget.open)}>
    <summary>Open for more data</summary>
    <p>More data</p>
  </details>
);

// With explicit typing: works properly
const TestComponent2 = () => (
  <details open={false} onToggle={(e: React.SyntheticEvent<HTMLDetailsElement>) => console.log('is now open?', e.currentTarget.open)}>
    <summary>Open for more data</summary>
    <p>More data</p>
  </details>
);
```

The tags I've identified with this behaviour are `<blockquote>`, `<del>`, `<details>`, `<meter>`, `<output>`, `<time>`.

They could be typed as

```tsx
interface IntrinsicElements {
  blockquote: React.DetailedHTMLProps<React.BlockquoteHTMLAttributes<HTMLQuoteElement>, HTMLQuoteElement>;
  del: React.DetailedHTMLProps<React.DelHTMLAttributes<HTMLModElement>, HTMLModElement>;
  details: React.DetailedHTMLProps<React.DetailsHTMLAttributes<HTMLDetailsElement>, HTMLDetailsElement>;
  meter: React.DetailedHTMLProps<React.MeterHTMLAttributes<HTMLMeterElement>, HTMLMeterElement>;
  output: React.DetailedHTMLProps<React.OutputHTMLAttributes<HTMLOutputElement>, HTMLOutputElement>;
  time: React.DetailedHTMLProps<React.TimeHTMLAttributes<HTMLTimeElement>, HTMLElement>;
}
```

The following also have this behaviour, but their own properties are marked as deprecated and/or they are marked as experimental themselves: `<caption>`, `<menu>`, `<picture>`.

Finally, `<keygen>` also has this behaviour, but is itself deprecated, so it should probably stay this way, I think.

[a comment by an automated account is not shown]

**Comment by eps1lon** [comment]
Looks like we should improve here. Do you want to work on a PR updating the used interfaces?

**Reply by sonicolasj** [reply]
Will do!

**Reply by sonicolasj** [reply, marked as answer]
I created a [pull-request](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/59274)!

I'm not sure if the tests are good enough (ie: they pollute the global scope), feel free to point out how to improve them!
